tabs.js

```diff
@@ -1,101 +1,281 @@
-        this.tabs.forEach((tab, index) => {
-            tab.classList.toggle('active', index === activeIndex);
-        });
+    _allPanels() {
+        return Array.from(this.querySelectorAll('iobio-tab-panel'));
+    }
+
+    _allTabs() {
+        return Array.from(this.querySelectorAll('iobio-tab'));
+    }
 
-        this.slottedElements.forEach((element, index) => {
-            element.classList.toggle('hidden', index == activeIndex ? false : true);
+    _panelForTab(tab) {
+        const panelId = tab.getAttribute('aria-controls');
+        return this.querySelector(`#${panelId}`);
+    }
+
+    reset() {
+        const tabs = this._allTabs();
+        const panels = this._allPanels();
+
+        tabs.forEach(tab => tab.selected = false);
+        panels.forEach(panel => panel.classList.add('hidden-panel'));
+    }
+
+    _selectTab(newTab) {
+        // Deselect all tabs and hide all panels.
+        this.reset();
+
+        // Get the panel that the `newTab` is associated with.
+        const newPanel = this._panelForTab(newTab);
+        // If that panel doesn’t exist, abort.
+        if (!newPanel)
+            throw new Error(`No panel with id ${newPanelId}`);
+        newTab.selected = true;
+        newPanel.classList.remove('hidden-panel');
+        newTab.focus();
+    }
+
+    _onClick(event) {
+        // If the click was not targeted on a tab element itself,
+        // it was a click inside the a panel or on empty space. Nothing to do.
+        if (event.target.getAttribute('role') !== 'tab')
+            return;
+        // If it was on a tab element, select that tab.
+        this._selectTab(event.target);
+    }
+
+    _addSeparators() {
+        const tabs = this.querySelectorAll('iobio-tab');
+        tabs.forEach((tab, index) => {
+            // Skip the last tab
+            if (index < tabs.length - 1) {
+                const separator = document.createElement('span');
+                separator.textContent = '|';
+                separator.style.margin = '0 10px';
+                separator.style.color = 'grey';
+                tab.appendChild(separator);
+            }
         });
     }
 }
-
 customElements.define('iobio-tabs', Tabs);
-export {Tabs};
\ No newline at end of file
+
+let TabCounter = 0;
+class Tab extends HTMLElement {
+    static get observedAttributes() {
+        return ['selected'];
+    }
+
+    constructor() {
+        super();
+    }
+
+    connectedCallback() {
+        // If this is executed, JavaScript is working and the element
+        // changes its role to `tab`.
+        this.setAttribute('role', 'tab');
+        if (!this.id)
+            this.id = `iobio-tab-generated-${TabCounter++}`;
+
+        // Set a well-defined initial state.
+        this.setAttribute('aria-selected', 'false');
+        this.setAttribute('tabindex', -1);
+
+        // Find the iobio-info-button and use its label as the tab name
+        const infoButton = this.querySelector('iobio-info-button');
```

Comment:
This breaks the abstraction as well. The tabs should never be aware of what type of element is inside them. The positioning logic needs to happen inside the info button element.